crates/wasm-smith/src/core.rs

```diff
@@ -509,6 +562,194 @@ impl Module {
+    ) -> Result<()> {
+        // First, parse the module-by-example to collect the types and imports.
+        let mut available_types = Vec::new();
+        let mut available_imports = Vec::new();
+        for payload in wasmparser::Parser::new(0).parse_all(example_module) {
+            match payload.expect("could not parse available import payload") {
+                wasmparser::Payload::TypeSection(mut type_reader) => {
+                    for _ in 0..type_reader.get_count() {
+                        let ty = type_reader.read().expect("could not parse type section");
+                        available_types.push((ty, None));
+                    }
+                }
+                wasmparser::Payload::ImportSection(mut import_reader) => {
+                    for _ in 0..import_reader.get_count() {
+                        let im = import_reader.read().expect("could not read import");
+                        // We can immediately filter whether this is an import we want to
+                        // use.
+                        let use_import = u.arbitrary().unwrap_or(false);
+                        if !use_import {
+                            continue;
+                        }
+                        available_imports.push(im);
+                    }
+                }
+                _ => {}
+            }
+        }
+
+        // In this function we need to place imported function/tag types in the types section and
+        // generate import entries (which refer to said types) at the same time.
+        let max_types = self.config.max_types();
+        let mut new_imports = Vec::with_capacity(available_imports.len());
+        let first_type_index = self.types.len();
+        let mut new_types = Vec::<Type>::new();
+
+        let convert_type = |parsed_type: wasmparser::Type| match parsed_type {
+            wasmparser::Type::I32 => ValType::I32,
+            wasmparser::Type::I64 => ValType::I64,
+            wasmparser::Type::F32 => ValType::F32,
+            wasmparser::Type::F64 => ValType::F64,
+            wasmparser::Type::V128 => ValType::V128,
+            wasmparser::Type::FuncRef => ValType::FuncRef,
+            wasmparser::Type::ExternRef => ValType::ExternRef,
+        };
+        let mut make_func_type = |parsed_sig_idx: u32| {
+            let serialized_sig_idx = match available_types.get_mut(parsed_sig_idx as usize) {
+                None => panic!("signature index refers to a type out of bounds"),
+                Some((_, Some(idx))) => *idx as usize,
+                Some((wasmparser::TypeDef::Func(func_type), index_store)) => {
+                    let new_index = first_type_index + new_types.len();
+                    if new_index < max_types {
+                        let func_type = Rc::new(FuncType {
+                            params: func_type.params.iter().map(|t| convert_type(*t)).collect(),
+                            results: func_type.returns.iter().map(|t| convert_type(*t)).collect(),
```

Comment:
Nit to self: we should disallow function types with multiple results without the multi_value feature.